modules/layouts/LayoutCS.chpl

```diff
@@ -0,0 +1,656 @@
+
+    // TODO: Allow zippering sparse domains of equal length some day...
+    if (followThisDom != this) then
+      halt("Sparse domains can't be zippered with anything other than themselves and their arrays (CS layout)");
+
+    // This loop is identical to the serial iterator, except for the iteration
+    // space and finding the initial 'cursorRow'.
+    var cursor = _private_findStart(startIx);
+    if debugCS then
+      writeln("CSDom follower: ", startIx, "..", endIx,
+              "  startIdx(", cursor, ")=", startIdx(cursor));
+
+    for i in startIx..endIx {
+      while (startIdx(cursor+1) <= i) do cursor+= 1;
+      if this.row then
+        yield (cursor, idx(i));
+      else
+        yield (idx(i), cursor);
+    }
+  }
+
+  iter these(param tag: iterKind, followThis) where tag == iterKind.follower {
+    compilerError("Sparse iterators can't yet be zippered with others (CS layout)");
+    yield 0;    // Dummy.
+  }
+
+  // Helper: find 'ix' s.t. startIdx(ix) <= startIx < startIdx(ix+1)
+  // or a number at most 'approx' smaller than that.
+  // There MUST exist a solution within low..high.
+  proc _private_findStart(startIx) {
+    return _private_findStart(startIx, dom.low, dom.high);
+  }
+  proc _private_findStart(startIx, low, high) {
+    var approx = 2; // Indicates when to switch to linear search.
+                    // This number could be tuned for performance.
+    // simple binary search (should be fewer comparisons than binarySearch())
+    var l = low, h = high;
+    while h > l + approx {
+      var m = (h + l) / 2;
+      if startIdx(m) <= startIx then l = m; else h = m;
+    }
+    var hh = min(l+approx, high);
+
+    if boundsChecking then {
+      assert(startIdx(l) <= startIx, "CSDom.findStart-1");
+      assert(startIx < startIdx(hh), "CSDom.findStart-2");
+      // The second assertion will fail if there is NO solution
+      // within low..high.  For performance,
+      // ensuring existence of a solution is caller responsibility.
+    }
+    return l;
+  }
+
+  proc stopIdx(i) {
+    return startIdx(i+1)-1;
+  }
+
+  proc find(ind: rank*idxType) {
+    use Search;
+
+    const (row, col) = ind;
+
+    var ret: (bool, idxType);
+    if this.row then
+      ret = binarySearch(idx, col, lo=startIdx(row), hi=stopIdx(row));
+    else
+      ret = binarySearch(idx, row, lo=startIdx(col), hi=stopIdx(col));
+
+    return ret;
+  }
+
+  proc dsiMember(ind: rank*idxType) {
+    if parentDom.member(ind) {
+      const (found, loc) = find(ind);
+      return found;
+    }
+    return false;
+  }
+
+  proc dsiFirst {
+    if nnz == 0 then return (parentDom.low) - (1,1);
+    const _low = nnzDom.low;
+    for i in dom do
+      if startIdx[i] > _low then
+        return (i-1, idx[idx.domain.low]);
+
+    halt("Something went wrong in dsiFirst");
+    return (0, 0);
+  }
+
+  proc dsiLast {
+    if nnz == 0 then return (parentDom.low) - (1,1);
+    const _low = nnzDom.low;
+    var _lastRow = parentDom.low[1] - 1;
```

Comment:
Good catch. Same with `dsiFirst`. Unfortunately this method does not produce the correct result on master today, so I need to do some debugging. Working on a fix..